File: lib/native_is_for_everyone/distro_managers/zypper.dart
import 'package:process_run/process_run.dart';
import 'package:xpm/native_is_for_everyone/models/native_package.dart';
import 'package:xpm/native_is_for_everyone/native_package_manager.dart';
import 'package:xpm/os/executable.dart';

/// Package manager adapter for openSUSE/SLES systems using Zypper
class ZypperPackageManager extends NativePackageManager {
  @override
  Future<List<NativePackage>> search(String name, {int? limit}) async {
    final shell = Shell(verbose: false);
    final result = await shell.run('zypper --non-interactive search "$name" 2>/dev/null');

    if (result.first.exitCode != 0) {
      return [];
    }

    final packages = _parseSearchOutput(result.first.stdout.toString());

    if (limit != null) {
      return packages.take(limit).toList();
    }

    return packages;
  }

  List<NativePackage> _parseSearchOutput(String output) {
    final packages = <NativePackage>[];
    final lines = output.split('\n');

    // Skip header lines (zypper has a table format with headers)
    bool headerPassed = false;

    for (final line in lines) {
      if (line.isEmpty) continue;

      // Skip separator lines
      if (line.startsWith('--') || line.startsWith('S ') || line.contains('| Name')) {
        headerPassed = true;
        continue;
      }

      if (!headerPassed) continue;

      // Zypper output format: "S | Name | Summary | Type"
      // or "i | package-name | Description | package"
      final parts = line.split('|');
      if (parts.length < 3) continue;

      final name = parts[1].trim();
      final description = parts[2].trim();

      if (name.isEmpty) continue;

      packages.add(NativePackage(
        name: name,
        description: description,
        repo: 'opensuse',
      ));
    }

    return packages;
  }

  @override
  Future<void> install(String name) async {
    final sudo = await Executable('sudo').find();
    if (sudo == null) {
      throw Exception('sudo not found');
    }
    final shell = Shell();
    await shell.run('$sudo zypper --non-interactive install "$name"');
  }

  @override
  Future<bool> isInstalled(String name) async {
    final shell = Shell(verbose: false);
    final result = await shell.run('rpm -q "$name" 2>/dev/null');
    return result.first.exitCode == 0;
  }

  @override
  Future<NativePackage?> get(String name) async {
    final shell = Shell(verbose: false);
    final result = await shell.run('zypper --non-interactive info "$name" 2>/dev/null');

    if (result.first.exitCode != 0) {
      return null;
    }

    return _parsePackageInfo(name, result.first.stdout.toString());
  }

  NativePackage _parsePackageInfo(String packageName, String output) {
    String? version;
    String? description;
    String? arch;
    String? repo;

    for (final line in output.split('\n')) {
      final colonIndex = line.indexOf(':');
      if (colonIndex == -1) continue;

      final field = line.substring(0, colonIndex).trim().toLowerCase();
      final value = line.substring(colonIndex + 1).trim();

      switch (field) {
        case 'version':
        case 'versão':
          version = value;
          break;
        case 'summary':
        case 'description':
        case 'descrição':
          description = value;
          break;
        case 'arch':
        case 'architecture':
        case 'arquitetura':
          arch = value;
          break;
        case 'repository':
        case 'repositório':
          repo = value;
          break;
      }
    }

    return NativePackage(
      name: packageName,
      version: version,
      description: description,
      arch: arch,
      repo: repo,
    );
  }
}

File: crates/xpm-core/src/native_pm/zypper.rs
//! Zypper package manager integration (openSUSE)

use super::{NativePackage, NativePackageManager};
use crate::os::executable::Executable;
use anyhow::{Context, Result};
use async_trait::async_trait;
use std::path::PathBuf;
use tokio::process::Command;

pub struct ZypperPackageManager {
    zypper_path: PathBuf,
    sudo_path: Option<PathBuf>,
}

impl ZypperPackageManager {
    pub async fn new() -> Self {
        let zypper_path = Executable::new("zypper")
            .find()
            .unwrap_or_else(|| PathBuf::from("/usr/bin/zypper"));

        let sudo_path = Executable::new("sudo").find();

        Self {
            zypper_path,
            sudo_path,
        }
    }

    async fn run_zypper(&self, args: &[&str]) -> Result<String> {
        let output = Command::new(&self.zypper_path)
            .args(args)
            .output()
            .await
            .context("Failed to run zypper command")?;

        if output.status.success() {
            Ok(String::from_utf8_lossy(&output.stdout).to_string())
        } else {
            let stderr = String::from_utf8_lossy(&output.stderr);
            anyhow::bail!("zypper command failed: {}", stderr)
        }
    }

    async fn run_zypper_sudo(&self, args: &[&str]) -> Result<String> {
        let mut cmd = if let Some(sudo) = &self.sudo_path {
            let mut cmd = Command::new(sudo);
            cmd.arg(&self.zypper_path);
            cmd
        } else {
            Command::new(&self.zypper_path)
        };

        let output = cmd
            .args(args)
            .output()
            .await
            .context("Failed to run zypper command")?;

        if output.status.success() {
            Ok(String::from_utf8_lossy(&output.stdout).to_string())
        } else {
            let stderr = String::from_utf8_lossy(&output.stderr);
            anyhow::bail!("zypper command failed: {}", stderr)
        }
    }

    fn parse_search_output(&self, output: &str) -> Vec<NativePackage> {
        let mut packages = Vec::new();

        for line in output.lines() {
            if line.trim().is_empty()
                || line.starts_with("Loading")
                || line.starts_with("S ")
                || line.starts_with("--")
            {
                continue;
            }

            let parts: Vec<&str> = line.split('|').map(|s| s.trim()).collect();
            if parts.len() >= 3 {
                let status = parts.first().unwrap_or(&"");
                let name = parts.get(1).unwrap_or(&"").to_string();
                let version = parts.get(3).map(|s| s.to_string());

                if name.is_empty() {
                    continue;
                }

                packages.push(NativePackage {
                    name,
                    version,
                    description: parts.get(2).map(|s| s.to_string()),
                    arch: parts.get(4).map(|s| s.to_string()),
                    repo: parts.get(5).map(|s| s.to_string()),
                    popularity: None,
                    installed: status.contains('i'),
                });
            }
        }

        packages
    }
}

#[async_trait]
impl NativePackageManager for ZypperPackageManager {
    fn name(&self) -> &str {
        "zypper"
    }

    async fn search(&self, query: &str, limit: Option<usize>) -> Result<Vec<NativePackage>> {
        let output = self.run_zypper(&["search", query]).await?;
        let mut packages = self.parse_search_output(&output);

        if let Some(limit) = limit {
            packages.truncate(limit);
        }

        Ok(packages)
    }

    async fn install(&self, name: &str) -> Result<()> {
        self.run_zypper_sudo(&["install", "-y", name]).await?;
        Ok(())
    }

    async fn remove(&self, name: &str) -> Result<()> {
        self.run_zypper_sudo(&["remove", "-y", name]).await?;
        Ok(())
    }

    async fn is_installed(&self, name: &str) -> Result<bool> {
        let output = Command::new("rpm").args(["-q", name]).output().await?;

        Ok(output.status.success())
    }

    async fn get(&self, name: &str) -> Result<Option<NativePackage>> {
        let output = self.run_zypper(&["info", name]).await;

        match output {
            Ok(output) => {
                let mut pkg = NativePackage::new(name);

                for line in output.lines() {
                    if let Some((key, value)) = line.split_once(':') {
                        let value = value.trim();
                        match key.trim() {
                            "Version" => pkg.version = Some(value.to_string()),
                            "Summary" | "Description" => pkg.description = Some(value.to_string()),
                            "Arch" => pkg.arch = Some(value.to_string()),
                            _ => {}
                        }
                    }
                }

                pkg.installed = self.is_installed(name).await.unwrap_or(false);
                Ok(Some(pkg))
            }
            Err(_) => Ok(None),
        }
    }

    async fn update_db(&self) -> Result<()> {
        self.run_zypper_sudo(&["refresh"]).await?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_search_output() {
        let pm = ZypperPackageManager {
            zypper_path: PathBuf::from("/usr/bin/zypper"),
            sudo_path: None,
        };

        let output = "S  | Name       | Summary          | Type   \n\
                      ---+------------+------------------+--------\n\
                      i  | vim        | Vi IMproved      | package\n\
                         | vim-data   | Vim data files   | package";

        let packages = pm.parse_search_output(output);

        assert_eq!(packages.len(), 2);
        assert_eq!(packages[0].name, "vim");
        assert!(packages[0].installed);
        assert_eq!(packages[1].name, "vim-data");
        assert!(!packages[1].installed);
    }

    #[test]
    fn test_name() {
        let pm = ZypperPackageManager {
            zypper_path: PathBuf::from("/usr/bin/zypper"),
            sudo_path: None,
        };
        assert_eq!(pm.name(), "zypper");
    }
}
